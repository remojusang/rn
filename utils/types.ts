export type RootStackParamList = {
  Tabs: undefined;
  Home: undefined;
  Login: { signedEmail: string } | undefined;
  Sign: undefined;
  Detail: PersonType;
  Camera: undefined;
  Profile: undefined;
};

export type PersonType = {
  name: string;
  age: number;
  gender: 'male' | 'female';
};

export type AlbumImageType = {
  uri: string;
};
