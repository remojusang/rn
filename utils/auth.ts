import auth from '@react-native-firebase/auth';

interface ISign {
  email: string;
  password: string;
}

function signIn({ email, password }: ISign) {
  return auth().signInWithEmailAndPassword(email, password);
}

function signUp({ email, password }: ISign) {
  return auth().createUserWithEmailAndPassword(email, password);
}

function subscribeAuth(callback: any) {
  return auth().onAuthStateChanged(callback);
}

function signOut() {
  return auth().signOut();
}

export { signIn, signUp, subscribeAuth, signOut };
