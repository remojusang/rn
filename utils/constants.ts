type RegexType = {
  email: RegExp;
  password: RegExp;
  phone: RegExp;
  passwordCheck?: RegExp; // TS에러 방지
};

const REGEX: RegexType = {
  email: /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/,
  password: /^[a-zA-Z0-9]{8,16}$/,
  phone: /^\d{3}-\d{3,4}-\d{4}$/,
} as const;

type IndexStringObject = { [index: string]: string };

const FORM_ERR_MSG: IndexStringObject = {
  email: '올바른 이메일 패턴이 아닙니다.',
  password: '8~16자 영문 대소문자, 숫자를 사용하세요.',
  required: '해당란을 입력해주세요.',
  passwordCheck: '비밀번호가 일치하지 않습니다.',
  phone: '올바른 휴대폰 패턴이 아닙니다.',
} as const;

type accessHintType = {
  LOADER: 'loader';
  PW: 'password';
  PW_CHECK: 'password_check';
};

const ACCESS_HINT: accessHintType = {
  LOADER: 'loader',
  PW: 'password',
  PW_CHECK: 'password_check',
} as const;

const COUNTRIES = ['한글', 'English', '日本語'];

const countryToLocale: IndexStringObject = {
  한글: 'ko',
  English: 'en-US',
  日本語: 'jp',
} as const;

const CAMERA_ERRCODE: IndexStringObject = {
  camera_unavailable: 'camera not available on device',
  permission: 'Permission not satisfied',
} as const;

export {
  REGEX,
  FORM_ERR_MSG,
  ACCESS_HINT,
  COUNTRIES,
  countryToLocale,
  CAMERA_ERRCODE,
};
