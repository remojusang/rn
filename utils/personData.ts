import { PersonType } from './types';

const personData: PersonType[] = [
  {
    name: 'Kalid',
    age: 15,
    gender: 'male',
  },
  {
    name: 'Fomagran',
    age: 20,
    gender: 'male',
  },
  {
    name: 'May',
    age: 19,
    gender: 'female',
  },
  {
    name: 'Julia',
    age: 29,
    gender: 'female',
  },
  {
    name: 'Hunter',
    age: 18,
    gender: 'male',
  },
  {
    name: 'Jane',
    age: 28,
    gender: 'female',
  },
  {
    name: 'Kendrick',
    age: 34,
    gender: 'male',
  },
  {
    name: 'Kali',
    age: 25,
    gender: 'female',
  },
  {
    name: 'Jay',
    age: 37,
    gender: 'male',
  },
  {
    name: 'Mac',
    age: 31,
    gender: 'male',
  },
];

export default personData;
