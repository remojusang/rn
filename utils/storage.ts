import AsyncStorage from '@react-native-async-storage/async-storage';

const getStorageData = async (key: string) => {
  return JSON.parse((await AsyncStorage.getItem(key)) || 'null');
};

const setStorageData = async (key: string, value: any) => {
  await AsyncStorage.setItem(key, JSON.stringify(value));
};

const removeStorageData = async (key: string) =>
  await AsyncStorage.removeItem(key);

export { getStorageData, setStorageData, removeStorageData };
