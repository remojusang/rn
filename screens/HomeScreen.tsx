import React from 'react';
import type { NativeStackScreenProps } from '@react-navigation/native-stack';
import { RootStackParamList } from '../utils/types';
import ScreenLayout from '../components/ScreenLayout';
import PersonList from '../components/List';
import LangSelector from '../components/LangSelector';
import { View } from 'react-native';
import { PersonType } from '../utils/types';
import Person from '../components/ListItem/Person';
import { personDataState } from '../components/Atoms';
import FloatingBtn from '../components/FloatingBtn';
import { useRecoilState } from 'recoil';

type HomeProps = NativeStackScreenProps<RootStackParamList, 'Home'>;

function HomeScreen({}: HomeProps) {
  const personData = useRecoilState(personDataState);
  return (
    <ScreenLayout isLoading={false}>
      <View style={{ width: '100%', alignItems: 'flex-end' }}>
        <LangSelector />
      </View>
      <PersonList
        isRefreshing={false}
        onRefresh={() => {}}
        flatListConfig={{
          data: personData[0] as PersonType[],
          renderItem: props => <Person {...props} />,
          keyExtractor: item => item.name,
        }}
      />
      <FloatingBtn />
    </ScreenLayout>
  );
}

export default HomeScreen;
