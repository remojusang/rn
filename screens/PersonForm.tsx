import { useRef, useState, useEffect, useCallback } from 'react';
import FormLayout from '../components/FormLayout';
import type { NativeStackScreenProps } from '@react-navigation/native-stack';
import {
  TextInput,
  ActivityIndicator,
  TouchableOpacity,
  Text,
} from 'react-native';
import { RootStackParamList, PersonType } from '../utils/types';
import { useForm } from 'react-hook-form';
import { useSetRecoilState } from 'recoil';
import { personDataState } from '../components/Atoms';
import FlexGap from '../components/FlexGap';
import FormInput from '../components/FormInput';
import RadioBtns from '../components/RadioBtns';
import { useIntl } from 'react-intl';
import Icon from 'react-native-vector-icons/Ionicons';
import { ACCESS_HINT } from '../utils/constants';

type PersonFormProps = NativeStackScreenProps<
  RootStackParamList,
  'PersonForm'
>;

function PersonForm({ navigation }: PersonFormProps) {
  const [loading] = useState(false);
  const { formatMessage } = useIntl();
  const [gender, setGender] = useState<'male' | 'female'>('male');
  const { formState, handleSubmit, control } = useForm<
    Pick<PersonType, 'name' | 'age'>
  >({
    mode: 'onChange',
  });
  const ageRef = useRef<TextInput>(null);
  const setPersonData = useSetRecoilState(personDataState);
  const onValid = ({
    name,
    age,
  }: Pick<PersonType, 'name' | 'age'>) => {
    if (!gender) {
      return;
    }
    setPersonData(prev => [{ name, age, gender }, ...prev]);
    navigation.navigate('Home');
  };
  const HeaderRightLoader = useCallback(
    () => <ActivityIndicator color="#fff" />,
    [],
  );
  const HeaderRight = () => (
    <TouchableOpacity onPress={handleSubmit(onValid)}>
      <Icon name="add" size={35} color="#2196f3" />
    </TouchableOpacity>
  );

  useEffect(() => {
    navigation.setOptions({
      headerRight: loading ? HeaderRightLoader : HeaderRight,
      ...(loading && { headerLeft: () => null }),
    });
    //eslint-disable-next-line
  }, [loading]);
  return (
    <FormLayout>
      <FormInput
        name="name"
        errorMsg={formState.errors.name?.message}
        control={control}
        textInputConfig={{
          placeholder: 'enter your name...',
        }}
        onNext={() => ageRef.current?.focus()}
      />
      <FlexGap gapSize={10} />
      <FormInput
        name="age"
        constraintslabel={formatMessage({
          id: 'ageConstraintslabel',
        })}
        errorMsg={formState.errors.age?.message}
        control={control}
        textInputConfig={{
          placeholder: 'enter your age...',
          keyboardType: 'number-pad',
        }}
        inputRef={ageRef}
      />
      <FlexGap gapSize={10} />
      <RadioBtns
        labelId="genderLabel"
        testId="genderRadioBtn"
        valueArray={['male', 'female']}
        setter={setGender}
      />
      <TouchableOpacity
        onPress={handleSubmit(onValid)}
        accessibilityHint={ACCESS_HINT.ADD_BTN}>
        <Text>TestingBtn</Text>
      </TouchableOpacity>
    </FormLayout>
  );
}

export default PersonForm;
