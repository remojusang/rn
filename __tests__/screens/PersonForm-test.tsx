import {
  act,
  fireEvent,
  render,
  screen,
} from '@testing-library/react-native';
import HomeScreen from '../../screens/HomeScreen';
import PersonForm from '../../screens/PersonForm';
import ErrorBoundary from '../../components/MyErrorBoundary';
import { IntlProvider } from 'react-intl';
import koMsg from '../../lang/ko.json';
import { ACCESS_HINT } from '../../utils/constants';
import RN from 'react-native';
import '@testing-library/jest-native/extend-expect';
import { RecoilRoot } from 'recoil';
import { ReactTestInstance } from 'react-test-renderer';
// 에러 방지 => Couldn't find a navigation object. Is your component inside NavigationContainer?
// https://spin.atomicobject.com/2021/02/24/react-navigation-5-unit-testing-components/
jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: jest.fn(),
    }),
  };
});
jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

describe('PersonForm', () => {
  const TEST_PROPS: any = {
    route: jest.fn(),
    navigation: {
      navigate: jest.fn(),
      setOptions: jest.fn(),
    },
  };
  test('렌더링 테스트', () => {
    const renderedJson = render(
      <ErrorBoundary>
        <IntlProvider locale="ko" messages={koMsg}>
          <PersonForm {...TEST_PROPS} />
        </IntlProvider>
      </ErrorBoundary>,
    ).toJSON();
    expect(renderedJson).toMatchSnapshot();
    expect(renderedJson).toBeTruthy();
  });
  test('아이템 추가', async () => {
    const TEST_AGE = 33;

    const { rerender } = render(
      <RecoilRoot>
        <IntlProvider locale="ko" messages={koMsg}>
          <PersonForm {...TEST_PROPS} />
          <HomeScreen {...TEST_PROPS} />
        </IntlProvider>
      </RecoilRoot>,
    );
    // jest.spyOn(RN, 'FlatList', 'get').mockImplementation(() => {
    //   return RN.FlatList;
    // });

    await act(async () => {
      await fireEvent.changeText(
        screen.getByPlaceholderText('enter your name...'),
        'james',
      );
      await fireEvent.changeText(
        screen.getByPlaceholderText('enter your age...'),
        TEST_AGE,
      );
      // 'male' 라디오 버튼
      await fireEvent.press(
        screen.getByTestId('genderRadioBtnInputgenderRadioBtn|0'),
      );
      // Add 버튼
      await fireEvent.press(
        await screen.getByAccessibilityHint(ACCESS_HINT.ADD_BTN),
      );
    });
    rerender(
      <RecoilRoot>
        <IntlProvider locale="ko" messages={koMsg}>
          <PersonForm {...TEST_PROPS} />
          <HomeScreen {...TEST_PROPS} />
        </IntlProvider>
      </RecoilRoot>,
    );
    expect(await screen.queryByText('James')).toBeVisible();
    expect(
      await screen.queryByText(`${TEST_AGE} years old`),
    ).toBeVisible();
    expect(await screen.queryByText('male ♂')).toBeVisible();
  });
});
