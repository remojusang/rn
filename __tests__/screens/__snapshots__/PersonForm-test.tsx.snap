// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonForm 렌더링 테스트 1`] = `
<RCTSafeAreaView
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "margin": 20,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "marginVertical": 16,
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "black",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            이름
          </Text>
          <Text
            style={
              {
                "fontWeight": "300",
                "marginLeft": 20,
              }
            }
          />
        </View>
        <TextInput
          accessibilityHint=""
          autoCapitalize="none"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="enter your name..."
          style={
            {
              "backgroundColor": "white",
              "borderColor": "#000",
              "borderRadius": 6,
              "borderWidth": 1,
              "fontSize": 18,
              "padding": 10,
            }
          }
        />
      </View>
      <View
        style={
          {
            "marginBottom": 10,
          }
        }
      />
      <View
        collapsable={false}
        style={
          {
            "marginVertical": 16,
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "black",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            나이
          </Text>
          <Text
            style={
              {
                "fontWeight": "300",
                "marginLeft": 20,
              }
            }
          >
            숫자만 입력
          </Text>
        </View>
        <TextInput
          accessibilityHint=""
          autoCapitalize="none"
          keyboardType="number-pad"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="enter your age..."
          style={
            {
              "backgroundColor": "white",
              "borderColor": "#000",
              "borderRadius": 6,
              "borderWidth": 1,
              "fontSize": 18,
              "padding": 10,
            }
          }
        />
      </View>
      <View
        style={
          {
            "marginBottom": 10,
          }
        }
      />
      <Text
        style={
          {
            "color": "#000",
            "fontSize": 16,
            "fontWeight": "bold",
            "marginBottom": 10,
          }
        }
      >
        성별
      </Text>
      <View
        style={
          [
            undefined,
            undefined,
            false,
          ]
        }
      >
        <View>
          <View
            style={
              [
                {
                  "flexDirection": "row",
                  "marginBottom": 5,
                },
                undefined,
                false,
              ]
            }
          >
            <View>
              <View
                accessibilityLabel="radioButtonInput0"
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "borderColor": "#2196f3",
                    "borderRadius": 15,
                    "borderWidth": 3,
                    "height": 30,
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": 30,
                  }
                }
                testID="genderRadioBtnInputgenderRadioBtn|0"
              >
                <View
                  style={
                    [
                      {
                        "borderRadius": 10,
                      },
                      {
                        "backgroundColor": "#2196f3",
                        "height": 20,
                        "width": 20,
                      },
                      {
                        "borderRadius": 10,
                        "height": 20,
                        "width": 20,
                      },
                      {
                        "backgroundColor": "#2196f3",
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              accessibilityLabel="radioButtonLabel0"
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  undefined,
                  {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "flexDirection": "row",
                  },
                ]
              }
              testID="genderRadioBtnLabelgenderRadioBtn|0"
            >
              <Text
                style={
                  [
                    {
                      "lineHeight": 20,
                      "paddingLeft": 10,
                    },
                    false,
                    {
                      "color": "#000",
                    },
                    undefined,
                  ]
                }
              >
                남성
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View
            style={
              [
                {
                  "flexDirection": "row",
                  "marginBottom": 5,
                },
                undefined,
                false,
              ]
            }
          >
            <View>
              <View
                accessibilityLabel="radioButtonInput1"
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "borderColor": "#2196f3",
                    "borderRadius": 15,
                    "borderWidth": 3,
                    "height": 30,
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": 30,
                  }
                }
                testID="genderRadioBtnInputgenderRadioBtn|1"
              >
                <View
                  style={
                    [
                      {
                        "borderRadius": 10,
                      },
                      false,
                      false,
                      false,
                    ]
                  }
                />
              </View>
            </View>
            <View
              accessibilityLabel="radioButtonLabel1"
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  undefined,
                  {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "flexDirection": "row",
                  },
                ]
              }
              testID="genderRadioBtnLabelgenderRadioBtn|1"
            >
              <Text
                style={
                  [
                    {
                      "lineHeight": 20,
                      "paddingLeft": 10,
                    },
                    false,
                    {
                      "color": "#000",
                    },
                    undefined,
                  ]
                }
              >
                여성
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        accessibilityHint="addBtn"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Text>
          TestingBtn
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
